filters/auth/jwt_validation.go

```diff
@@ -0,0 +1,171 @@
+)
+
+const (
+	JwtValidationName = "jwtValidation"
+)
+
+type (
+	jwtValidationSpec struct {
+		options TokenintrospectionOptions
+	}
+
+	jwtValidationFilter struct {
+		jwksUri string
+	}
+)
+
+//the map of jwks keyfunctions stored per jwksUri
+var jwksMap map[string]*keyfunc.JWKs = make(map[string]*keyfunc.JWKs)
+
+func NewJwtValidationWithOptions(o TokenintrospectionOptions) filters.Spec {
+	return &jwtValidationSpec{
+		options: o,
+	}
+}
+
+func NewJwtValidation(timeout time.Duration) filters.Spec {
+	return NewJwtValidationWithOptions(TokenintrospectionOptions{
+		Timeout: timeout,
+		Tracer:  opentracing.NoopTracer{},
+	})
+}
+
+func JwtValidationWithOptions(
+	create func(time.Duration) filters.Spec,
+	o TokenintrospectionOptions,
+) filters.Spec {
+	s := create(o.Timeout)
+	ts, ok := s.(*jwtValidationSpec)
+	if !ok {
+		return s
+	}
+
+	ts.options = o
+	return ts
+}
+
+func (s *jwtValidationSpec) Name() string {
+	return JwtValidationName
+}
+
+func (s *jwtValidationSpec) CreateFilter(args []interface{}) (filters.Filter, error) {
+	sargs, err := getStrings(args)
+	if err != nil {
+		return nil, err
+	}
+	if len(sargs) < 1 {
+		return nil, filters.ErrInvalidFilterParameters
+	}
+
+	issuerURL := sargs[0]
+
+	cfg, err := getOpenIDConfig(issuerURL)
+	if err != nil {
+		return nil, err
+	}
+
+	f := &jwtValidationFilter{
+		jwksUri: cfg.JwksURI,
+	}
+
+	return f, nil
+}
+
+func getKeyFunction(url string) (jwks *keyfunc.JWKs, err error) {
+	if jwks, ok := jwksMap[url]; ok {
+		return jwks, nil
+	}
+
+	// Create the keyfunc options. Use an error handler that logs. Refresh the JWKs when a JWT signed by an unknown KID
+	// is found or at the specified interval. Rate limit these refreshes. Timeout the initial JWKs refresh request after
+	// 10 seconds. This timeout is also used to create the initial context.Context for keyfunc.Get.
+	refreshInterval := time.Hour
+	refreshRateLimit := time.Minute * 5
+	refreshTimeout := time.Second * 10
+	refreshUnknownKID := true
+	options := keyfunc.Options{
+		RefreshErrorHandler: func(err error) {
+			log.Errorf("There was an error with the jwt.KeyFunc\nError:%s\n", err.Error())
+		},
+		RefreshInterval:   &refreshInterval,
+		RefreshRateLimit:  &refreshRateLimit,
+		RefreshTimeout:    &refreshTimeout,
+		RefreshUnknownKID: &refreshUnknownKID,
+	}
+
+	jwks, err = keyfunc.Get(url, options)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get the JWKs from the given URL %s Error:%s", url, err.Error())
+	}
+	jwksMap[url] = jwks
+	return jwks, nil
+
+}
+
+func (f *jwtValidationFilter) Request(ctx filters.FilterContext) {
+	r := ctx.Request()
+
+	var info tokenContainer
+	infoTemp, ok := ctx.StateBag()[oidcClaimsCacheKey]
+	if !ok {
+		token, ok := getToken(r)
+		if !ok || token == "" {
+			unauthorized(ctx, "", missingToken, "", "")
+			return
+		}
+
+		jwks, err := getKeyFunction(f.jwksUri)
+		if err != nil {
+			log.Fatalf("Failed to get the JWKs from the given URL %s. Error:%s", f.jwksUri, err.Error())
```

Comment:
I am wondering should we rather pass debugInfo to the `unauthorized()` instead of logging error here. Also %v format specifier formats error without the need to call .Error()